### PR TITLE
outbound: Split TCP stack into dedicated modules

### PR DESCRIPTION
**Pull request by olix0r**
The main `outbound/src/lib.rs` is getting large and unwieldy. Before
making additional changes, this reorganizes outbound as follows:

* `endpoint` has been renamed to `target`. It includes target types,
  one of which is `Endpoint`.
* A new `tcp` module includes TCP-specific target aliases.
* The TCP connect stack is built in `tcp::connect`. The `PreventLoop`
  filter is now a private type within this module.
* The TCP balance stack is built in `tcp::balance`.
* The `resolve` module is no longer generic over its target types, now
  relying on the `Concrete` target type.
* `allow_discovery::AllowProfile` is now `tcp::AllowProfile`, as it is
  specific to the `tcp::Accept` target type.
* The `tcp::connect` stack is now generic over the endpoint's protocol
  (instead of hardcoding implementations for TcpEndpoint and
  HttpEndpoint). This means that separate concrete connect stacks must
  be used for each protocol.
* TCP tests have been moved under the `tcp` module.

There are no functional changes. Similar changes will follow for the HTTP
stacks.

### CODE REVIEW

File: linkerd/app/outbound/src/resolve.rs [hunk 1]
```diff
@@ -24,59 +24,69 @@ use std::{
...
+    type Request = Addr;
+
+    fn filter(&self, target: Concrete<P>) -> Result<Addr, Error> {
+        target.resolve.ok_or_else(|| discovery_rejected().into())
+    }
+}
+
 // === impl RecoverDefault ===
 
 type Resolution<R> =
```
Comment (hawkw) [inline linkerd/app/outbound/src/resolve.rs:92]: very minor style nit: i have a slight preference for having the `impl` blocks in the same order as the types are defined, not a big deal.

File: linkerd/app/outbound/src/target.rs [hunk 2]
```diff
@@ -79,22 +90,31 @@ impl Into<transport::labels::Key> for &'_ Accept {
...
-    fn from((protocol, l): (http::Version, TcpLogical)) -> Self {
+impl<P> From<(http::Version, Logical<P>)> for HttpLogical {
+    fn from(
+        (
+            protocol,
+            Logical {
+                orig_dst, profile, ..
+            },
+        ): (http::Version, Logical<P>),
+    ) -> Self {
```
Comment (hawkw) [inline linkerd/app/outbound/src/target.rs:113]: wow...

File: linkerd/app/outbound/src/tcp/connect.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        // Initiates mTLS if the target is configured with identity.
+        .push(tls::client::ConnectLayer::new(local_identity))
+        // Limits the time we wait for a connection to be established.
+        .push_timeout(config.timeout)
+        .push(metrics.transport.layer_connect())
+        .push_request_filter(PreventLoop { port: server_port })
+        .into_inner()
+}
+
+/// A connection policy that drops
```
Comment (hawkw) [inline linkerd/app/outbound/src/tcp/connect.rs:31]: drops what?

File: linkerd/app/outbound/src/test_util.rs [hunk 4]
```diff
@@ -1,31 +1,21 @@
...
 use linkerd2_app_core::{
-    config, exp_backoff, profiles,
-    proxy::http::h2,
-    svc::NewService,
-    transport::{listen, tls},
-    Error, IpMatch,
+    config, exp_backoff, profiles, proxy::http::h2, transport::listen, IpMatch,
 };
-use linkerd2_app_test as test_support;
+pub use linkerd2_app_test as support;
```
Comment (hawkw) [inline linkerd/app/outbound/src/test_util.rs:7]: nit/TIOLI: might be slightly nicer to just
```suggestion
pub use linkerd2_app_test::*;
```
 and then have the tests just use _this_ module as `support`? not a blocker.

**Review by hawkw — APPROVED** [review]
looks good to me; couple of minor nits i don't really care about

File: linkerd/app/outbound/src/test_util.rs (hunk 4 above)
Comment (olix0r) [reply linkerd/app/outbound/src/test_util.rs:7]: maybe in a followup. glob imports end up being pretty hairy to understand, and in this case it's easy for unintentional overlaps like `io() vs transport::io`

File: linkerd/app/outbound/src/resolve.rs (hunk 1 above)
Comment (olix0r) [reply linkerd/app/outbound/src/resolve.rs:92]: aren't they already? what needs to change?

Comment (hawkw) [reply linkerd/app/outbound/src/resolve.rs:92]: oh you know what? never mind. i can't read.

**Pull request merged**